### ISSUE TITLE
Bug: [indent] Maximum call stack size exceeded

### ISSUE THREAD

**Issue by IronGeek** [typescript-eslint/typescript-eslint#11455]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.tsx&code=MYewdgzgLgBAZiEMC8MAUBKFA%2BdAoGGAHmwMOIBMBLANxgHpTzLaGnjG8MBuIA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQJYDsATRfAF3TAG1JFpoB7aSAGjACYBdcMAXxB6A&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
const foo = () => (
  <>
    <div />
    <div />
  </>
);
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "indent": ["error", 2]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Should not resulted in `RangeError: Maximum call stack size exceeded`

### Actual Result

```
RangeError: Maximum call stack size exceeded
    at Map.get (<anonymous>)
    at s.getFirstTokenOfLine (https://typescript-eslint.io/sandbox/index.js:1:876916)
    at a.getDesiredIndent (https://typescript-eslint.io/sandbox/index.js:1:878206)
    at a.getDesiredIndent (https://typescript-eslint.io/sandbox/index.js:1:878173)
    at a.getDesiredIndent (https://typescript-eslint.io/sandbox/index.js:1:878173)
    at a.getDesiredIndent (https://typescript-eslint.io/sandbox/index.js:1:878173)
    at a.getDesiredIndent (https://typescript-eslint.io/sandbox/index.js:1:878173)
    at a.getDesiredIndent (https://typescript-eslint.io/sandbox/index.js:1:878173)
    at a.getDesiredIndent (https://typescript-eslint.io/sandbox/index.js:1:878173)
    at a.getDesiredIndent (https://typescript-eslint.io/sandbox/index.js:1:878173) 1:1 - 6:3
```

### Additional Info

I know this error happen with [`@stylistic/indent`](https://eslint.style/rules/indent) or the deprecated [indent](https://eslint.org/docs/latest/rules/indent) rule.
But the origin of the error seems to be from `@typescript-eslint/parser` with `typescript@5.9.*` (it works fine with `typescript@5.8.3`). 

Not sure if it should be fixed here or at ESLint Stylistic, but the ESLint Stylistic thinks this is an upstream issue. 
From the ESLint Stylistic issue thread on   https://github.com/eslint-stylistic/eslint-stylistic/issues/915 :

> [@ota-meshi](https://github.com/eslint-stylistic/eslint-stylistic/issues/915#issuecomment-3146066717):
> _It seems that the tokens returned by @typescript-eslint/parser differ depending on the TypeScript version.
In TypeScript 5.8, </> returns three tokens: <, /, and >,
but in TypeScript 5.9, </> returns two tokens: </ and >.
The standard ESLint parser returns three tokens. Therefore, I think this is a problem with @typescript-eslint/parser._

---

FWIW, I tested the same rule configuration without TypeScript [ESLint](https://eslint.org/play/#eyJ0ZXh0IjoiY29uc3QgZm9vID0gKCkgPT4gKFxuICA8PlxuICAgIDxkaXYgLz5cbiAgICA8ZGl2IC8+XG4gIDwvPlxuKTsiLCJvcHRpb25zIjp7InJ1bGVzIjp7ImluZGVudCI6WyJlcnJvciIsMl0sInNlbWkiOlsiZXJyb3IiLCJhbHdheXMiXX0sImxhbmd1YWdlT3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwic291cmNlVHlwZSI6Im1vZHVsZSJ9fX0=) [ESLint with Stylistic](https://eslint-online-playground.netlify.app/#eNqVUk1vgzAM/StWLmulfkg70g9V2rTrLrs1PURgUDpIoiRUrRD/fQ6BtmvFoRxwsN/zs/NomLPpEs+iMiUujo4lLNXKeci1hg1MprDZwoQrgPU2vClm8gTL5491CNMVmzF0pVR+QX1yWcSesjLaemggw1wq/OhK0EJudQW8Zywjg7MVVz3B+UspnZdpRL7troll5MxNWRdSvRGFKzx3JNIQddnFq9Zkzz1BfANx8FKoohYFfhsvad8EmpgHMMI6tM95ejCtxBcKX1t8qNBzdOcEvK3xLt0O53Y2nJyubYo/F4MJLV7prC6Rs1gdUHGne4nr2v+BlsgBRp1uFyNVhspzlsCebtZabTmbwfuBrjvwDp1HRqS/tD65oxX50wlx5mmsQLwN1glxluHpEw2GzqlEFzD9bJzteveOXZqzUnh0pN/POLg7Uhxz9EX4fFR9lOFfZpDFr1IoPMgECv0XLWv/ALZ+IqI=) and it works fine.

**Comment by bradzacher** [comment]
This was an expected change with the new version of TypeScript - they merged their underlying tokens.

Tokens aren't something specced as part of ESLint/ESTree.  But if it's causing issues we can consider aligning with eslint.

**Comment by kirkwaiblinger** [comment]
Linking for visibility - the TS team is considering reverting the change that caused this: https://github.com/microsoft/TypeScript/issues/62188#issuecomment-3159386601

**Comment by jakebailey** [comment]
I'm on the fence about it, because it really was a bug that we were emitting `<` `/` `>`. and not `</` `>` in this specific API case; `</Fragment>` will give `</` `Fragment` `>` IIRC so...

**Comment by JoshuaKGoldberg** [comment]
IMO we should treat this as a bug and fix it. The TS change isn't reverted, and we still support the TS version(s) with the new behavior.

Aside: it's nice to see that "only" ~15-20 folks have 👍d this. It used to be `indent` rule complaints received many dozens of upvotes. 🙂

**Comment by kirkwaiblinger** [comment]
FYI this was reverted in TS (https://github.com/microsoft/TypeScript/pull/62423) so the next version of TS should fix it but it sounds like it will eventually become an issue to be addressed again.

**Comment by JoshuaKGoldberg** [comment]
For now, closing out as resolved in TypeScript. But if someone sees this pop up again please do file a new issue. Thanks all!
